DREAM/Optimization/tsgParticleSwarm.cpp

```diff
@@ -0,0 +1,168 @@
+ * FROM OR ARISING OUT OF, IN WHOLE OR IN PART THE USE, STORAGE OR DISPOSAL OF THE SOFTWARE.
+ */
+
+#ifndef __TASMANIAN_PARTICLE_SWARM_SOLVER_CPP
+#define __TASMANIAN_PARTICLE_SWARM_SOLVER_CPP
+
+#include "tsgParticleSwarm.hpp"
+
+namespace TasOptimization {
+
+ParticleSwarmState::ParticleSwarmState() : num_iterations(0), num_dimensions(0), num_particles(0) {}
+
+void ParticleSwarmState::setParticlePositions(std::vector<double> &pp) {
+    if (pp.size() % num_dimensions != 0) {
+        throw std::runtime_error("Size of particle positions (" + std::to_string(pp.size()) + ") is not a multiple of the " +
+                                 "number of dimensions (" + std::to_string(num_dimensions) + ")");
+    }
+    particle_positions = pp;
+}
+
+
+void ParticleSwarmState::setParticleVelocities(std::vector<double> &pv) {
+    if (pv.size() % num_dimensions != 0) {
+        throw std::runtime_error("Size of particle velocities (" + std::to_string(pv.size()) + ") is not a multiple of the " +
+                                 "number of dimensions (" + std::to_string(num_dimensions) + ")");
+    }
+    particle_velocities = pv;
+}
+
+
+void ParticleSwarmState::setBestParticlePositions(std::vector<double> &bpp) {
+    if (bpp.size() % num_dimensions != 0) {
+        throw std::runtime_error("Size of best particle positions (" + std::to_string(bpp.size()) + ") is not a multiple of the " +
+                                 "number of dimensions (" + std::to_string(num_dimensions) + ")");
+    }
+    best_particle_positions = bpp;
+}
+
+void ParticleSwarmState::setBestPosition(std::vector<double> &bp) {
+    if (bp.size() != (size_t) num_dimensions) {
+        throw std::runtime_error("Size of best position (" + std::to_string(bp.size()) + ") is not equal to the number of " +
+                                 "dimensions (" + std::to_string(num_dimensions) + ")");
+    }
+    best_position = bp;
+}
+
+void ParticleSwarmState::addParticlesInsideBox(int box_num_particles, std::vector<double> &box_lower, std::vector<double> &box_upper,
+                                               std::function<double(void)> get_random01) {
+    if (box_lower.size() != box_upper.size())
+        throw std::runtime_error("The size of box lower bounds and box upper bounds do not match");
+    if (num_dimensions == 0) num_dimensions = box_lower.size();
+    num_particles += box_num_particles;
+
+    std::vector<double> box_particle_positions(num_dimensions * box_num_particles);
+    std::generate(box_particle_positions.begin(), box_particle_positions.end(), get_random01);
+    for (int i=0; i<box_num_particles; i++) {
+        for (int j=0; j<num_dimensions; j++) {
+            box_particle_positions[i * num_dimensions + j] *= (box_upper[i] - box_lower[i]);
+            box_particle_positions[i * num_dimensions + j] -= box_lower[i];
+        }
+    }
+    std::copy(box_particle_positions.begin(), box_particle_positions.end(), std::back_inserter(particle_positions));
```

Comment:
Why adding more particles, it's OK to keep the number fixed.

Reply:
See my comment about ``num_particles`` above.